### PR TITLE
Dedup import scopes based on value equality instead of reference equality

### PR DESCRIPTION
**Pull request by tmat**
Bound usings are associated with InContainerBinders. Since binders are allocated as needed and only some instances are cached usings can be bound multiple times resulting in multiple instances of ImportChain. Portable PDB needs to compare them by value equality instead of reference equality in order to avoid non-deterministic output.

### CODE REVIEW

**Comment by tmat** [comment]
@dotnet/roslyn-compiler Please review.

**Comment by tmat** [comment]
@dotnet-bot retest prtest/win/dbg/unit64 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit64/4908/
// Retest reason:

File: src/Compilers/CSharp/Test/Emit/Emit/DeterministicTests.cs [hunk 1]
```diff
@@ -37,7 +38,7 @@ private ImmutableArray<byte> EmitDeterministic(string source, Platform platform,
         {
             var options = (optimize ? TestOptions.ReleaseExe : TestOptions.DebugExe).WithPlatform(platform).WithDeterministic(true);
 
-            var compilation = CreateCompilation(source, assemblyName: "DeterminismTest", references: new[] { MscorlibRef }, options: options);
+            var compilation = CreateCompilation(source, assemblyName: "DeterminismTest", references: new[] { MscorlibRef, SystemCoreRef, CSharpRef }, options: options);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/DeterministicTests.cs]: Why is this change needed?

Comment (tmat) [inline src/Compilers/CSharp/Test/Emit/Emit/DeterministicTests.cs]: So that dynamic works.

File: src/Compilers/Core/Portable/PEWriter/MetadataWriter.PortablePdb.cs [hunk 2]
```diff
@@ -78,7 +79,28 @@ private struct CustomDebugInformationRow
...
+
+            public bool Equals(IImportScope x, IImportScope y)
+            {
+                return (object)x == y ||
+                       x != null && y != null && Equals(x.Parent, y.Parent) && x.GetUsedNamespaces().SequenceEqual(y.GetUsedNamespaces());
+            }
+
+            public int GetHashCode(IImportScope obj)
+            {
+                return Hash.CombineValues(obj.GetUsedNamespaces(), obj.Parent?.GetHashCode() ?? 0);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/MetadataWriter.PortablePdb.cs]: It looks like the second argument is maxItemsToHash rather than another value to combine with. Was it your intent?

Comment (tmat) [inline src/Compilers/Core/Portable/PEWriter/MetadataWriter.PortablePdb.cs]: Good catch. Will fix.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PEWriter/MetadataWriter.PortablePdb.cs]: I think we need to change the name of the method(s) so that it would hard to make similar mistake in the future.

**Comment by cston** [comment]
LGTM

**Comment by gafter** [comment]
LGTM
